Create Section Label

andrew-nowak opened the pull request:
Resolves #437 

**Overall change:** Implements the ~~section divider~~ section label component

**Code changes:**

- A second attempt after #445 at implementing without using flexbox.

New:
  - ~~Exports two components; a `SectionLabel` and a `SectionLabelWithoutBar` to accommodate the two UX designs~~

Even newer:
  - Control whether you see a bar by passing in a prop `bar={false}`
  - Receive a prop to use as `id` for the label, to allow using `aria-labelledby` in the container.
  - Not providing a strapline counts as an error

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval

andrew-nowak committed:
First implementation of section divider

andrew-nowak committed:
section divider: require children to be string

andrew-nowak committed:
minor improvements

andrew-nowak committed:
changelog

andrew-nowak committed:
Try a non-flexbox approach

andrew-nowak committed:
continue non-flexbox approach

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,87 @@
+
+const TOP_MARGIN_PX = 1;
+
+const calcTop = ({ inline, script }) =>
```
I know eslint is making it this way, but is there anyway to appease it, while making it look nice and easy to read? :P

andrew-nowak replied:
~~How does 722d3ab look?~~

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,87 @@
+  font-family: ${GEL_FF_REITH_SANS};
+  display: inline-block;
+  position: relative;
+  margin: ${TOP_MARGIN_PX}px 0 0 0;
```
This can probably just be this right? 
```suggestion
  margin-top: ${TOP_MARGIN_PX}px;
```

andrew-nowak replied:
Can it? My thinking was to force the default margins to 0 so that margins added to the Wrapper or through the container can control the placement

dr3 replied:
Ah good point

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,87 @@
+  position: relative;
+  margin: ${TOP_MARGIN_PX}px 0 0 0;
+
+  html:not([dir='rtl']) & {
```
Whats the browser support for this like? Im not sure if I prefer this, vs passing `dir` as a prop to the component

dr3 replied:
If we do, We can probably reduce the complexity here if we just account for the `html[dir='rtl']`

Have it default to `padding: 0 ${GEL_SPACING} 0 0;` then override if `html[dir='rtl']` to `padding: 0 0 0 ${GEL_SPACING};`

dr3 replied:
chatting to @pjlee11 in #si_repo_psammead on slack, we should pass `dir` as a prop and toggle on `rtl` as a value

andrew-nowak replied:
Yep, that sounds good. I don't have any attachment to this method, I was just trying to make the rtl toggle on storybook work (which may need some thought about how that would work with dir as a prop)

dr3 reviewed (changes requested):
![](https://media.giphy.com/media/l46CkATpdyLwLI7vi/giphy.gif)

Amazing work, really really nice, just a couple things

andrew-nowak committed:
refactor confusing fn

andrew-nowak committed:
reupdate changelog

andrew-nowak committed:
divider text hangs only at >= 1280px

andrew-nowak committed:
top margin only when dangling

andrew-nowak commented:
~~Blocked on #456~~

andrew-nowak committed:
Merge remote-tracking branch 'origin/latest' into section-divider-nonflexbox

andrew-nowak committed:
Merge remote-tracking branch 'origin/latest' into section-divider-nonflexbox

andrew-nowak committed:
dir is passed as prop, not read from root element

andrew-nowak committed:
Merge branch 'latest' into section-divider-nonflexbox

andrew-nowak committed:
README updates

andrew-nowak committed:
unregister storybook viewport addon

andrew-nowak committed:
script prop is optional

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,113 @@
+SectionDivider.defaultProps = {
+  children: '',
+  dir: 'ltr',
+  script: {},
```
```suggestion
  script: null,
```

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,113 @@
+);
+
+SectionDivider.defaultProps = {
+  children: '',
```
```suggestion
  children: null,
```

dr3 commented on packages/components/psammead-section-divider/src/index.test.jsx:
```diff
@@ -0,0 +1,26 @@
+
+describe('SectionDivider', () => {
+  describe('With title', () => {
+    shouldMatchSnapshot(
```
This tests that `ltr` is the default, can we have one that passes it in explicitly please :)

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,113 @@
+  font-family: ${GEL_FF_REITH_SANS};
+  display: inline-block;
+  position: relative;
+  margin: ${MARGIN_TOP_PX}px 0 0 0;
```
Can this just use `margin-top`? Also we should use `rem` here :)

andrew-nowak replied:
> Can this just use `margin-top`?

No, this is to override the browser's default margins for an h2. I'll add a comment explaining why

> Also we should use `rem` here :)

:+1:

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,113 @@
+import { oneOf, shape, string } from 'prop-types';
+import { scriptPropType } from '@bbc/gel-foundations/prop-types';
+import {
+  GEL_GROUP_3_SCREEN_WIDTH_MIN, // 600px
```
Im not a big fan of these comments as they technically could go out of sync with the package they reference. Can they be removed please? Lines 6->22

dr3 commented on packages/components/psammead-section-divider/package.json:
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "@bbc/psammead-section-divider",
+  "version": "0.1.0",
+  "description": "React styled component for a section-divider",
```
```suggestion
  "description": "React styled component for a section divider",
```

dr3 commented on packages/components/psammead-section-divider/src/index.jsx:
```diff
@@ -0,0 +1,113 @@
+    position: absolute;
+    height: 1px;
+    border: none;
+    background: ${C_CLOUD_LIGHT};
```
In zeplin its using the colour `#aeaeb5`, is this wrong?

andrew-nowak replied:
Caught up with UX, the colour has changed from C_CLOUD_LIGHT to C_PEBBLE (`#aeaeb5`)... except C_PEBBLE doesn't exist in `psammead-styles`, so I raised #501

dr3 reviewed (changes requested):
Looking good, Im not sure about the behaviour at 1280px, we may need to chat over how its gunna work in simorgh with the story promo list as Im not convinced it can work like it is atm

dr3 commented:
![image](https://user-images.githubusercontent.com/11341355/57286781-f561ef00-70ad-11e9-8aa6-83bd3a006234.png)

For now, we can remove the code for the 1280px breakpoint, as it will involve additional logic to escape the grid column/component magic

sareh commented on packages/components/psammead-section-divider/README.md:
```diff
@@ -0,0 +1,60 @@
+
+### Accessibility notes
+
+Although this component has the appearance of a horizontal rule, it does not use an `<hr>` tag, and therefore does not have the associated semantic meaning.
```
It's great that you've added this note about the semantics of this component. 👍

sareh commented on packages/components/psammead-section-divider/README.md:
```diff
@@ -0,0 +1,60 @@
+import SectionDivider from '@bbc/psammead-section-divider';
+import { latin } from '@bbc/gel-foundations/scripts';
+
+const WrappingComponent = () => <SectionDivider script={latin} dir="ltr">Text here</SectionDivider>;
```
Is the recommended usage of this component to be with a heading element? for example, with an `<h2>`? If so, it would be good for the sample usage be using this. e.g.

```jsx
import SectionDivider from '@bbc/psammead-section-divider';
import { SubHeading } from '@bbc/psammead-headings';
import { latin } from '@bbc/gel-foundations/scripts';

const Wrapper = () => (

const WrappingComponent = () => (
  <SectionDivider script={latin}>
    <SubHeading script={latin}>Some subheadline</SubHeading>
  </SectionDivider>
);
```

Also, is it necessary for there to be a `dir="ltr"` on this SectionDivider element? I believe just having it further up the page should be fine.  The reason I'm noting this, is because often people copy and paste the Readme example into their code, so it's good to make sure we have examples that people can clearly use immediately.

andrew-nowak replied:
> Is the recommended usage of this component to be with a heading element? 

No, it isn't. This component puts the children it receives (and as the prop-types make clear, it should only be passed a string) in its own `<h2>`, and intentionally not the SubHeading component, as it uses Double Pica sizing, rather than Trafalgar as in the SubHeading. I will make this more clear on the readme.

> Also, is it necessary for there to be a `dir="ltr"` on this SectionDivider element? I believe just having it further up the page should be fine. The reason I'm noting this, is because often people copy and paste the Readme example into their code, so it's good to make sure we have examples that people can clearly use immediately.

No it's not required, but it is valid, the example would work, and IMO helps to remind someone who is copy-and-pasting that, unlike the majority of other components, the styling won't 'just work' for rtl text and whether the prop is being set needs to be considered.

sareh commented on packages/components/psammead-section-divider/README.md:
```diff
@@ -0,0 +1,60 @@
+
+## Description
+
+The `@bbc/psammead-section-divider` package exports a single Section Divider component.
```
This is a comment about the naming of this component within Psammead.  Can it please be renamed as `@bbc/psammead-divider`? This is because the use of `section-divider` appears to refer to the HTML semantic `<section>`, which is not the case here.

andrew-nowak replied:
I had a catch up with UX about this component yesterday, they are now referring to this component as a 'Section Label'. Would it be best to rename this component to align as closely as possible with their name? This would also have the issue that `section-label` might be referring to `<section>`, but `label` also feels very vague.

Commit:
Prefer `null` for default props

Co-Authored-By: aacn500 <10963046+aacn500@users.noreply.github.com>

Commit:
fix package description typo

Co-Authored-By: aacn500 <10963046+aacn500@users.noreply.github.com>

andrew-nowak committed:
Add snapshot test for explicit setting of text dir

andrew-nowak committed:
Prefer rems to pxs and explain setting margins to 0

andrew-nowak committed:
Remove constants value hints

andrew-nowak committed:
remove 1280px breakpoint "escaping the grid" behaviour

andrew-nowak committed:
clarify things on readme

andrew-nowak committed:
Margins for 600px breakpoint; non-bold font

andrew-nowak commented:
~~Blocked on #501 - see [this comment](https://github.com/bbc/psammead/pull/449#discussion_r281716950)~~
No longer blocked :)

amywalkerdev committed:
Merge branch 'latest' into section-divider-nonflexbox

andrew-nowak committed:
Use Pebble for bar colour

andrew-nowak committed:
Export section divider with and without a bar

andrew-nowak committed:
use alpha versioning

andrew-nowak committed:
renaming to section label

andrew-nowak committed:
Updates README

andrew-nowak committed:
Merge branch 'latest' into section-divider-nonflexbox

j-pendlebury reviewed (dismissed):
![](https://media.giphy.com/media/UVTJOUHNmy84TRne5m/giphy.gif)

andrew-nowak committed:
actually test the functionality

andrew-nowak committed:
actually test the functionality

andrew-nowak committed:
Using props to set the bar

pjlee11 commented on packages/components/psammead-section-label/src/index.jsx:
```diff
@@ -0,0 +1,147 @@
+};
+
+SectionLabel.propTypes = {
+  children: string,
```
Are `children` required?

andrew-nowak replied:
~~not in my implementation, I've handled the case where children aren't provided. My thinking was to handle the case where a strapline was missing from the data, and rendering the bar without the extra spacing for the missing text - see <https://github.com/bbc/psammead/pull/449/files#diff-a826825d0bceab40bf039858608d0224R56>~~

This is no longer the case, it's been confirmed that (at least for now) we will consider a missing strapline an error and won't handle that case

pjlee11 commented on packages/components/psammead-section-label/src/index.jsx:
```diff
@@ -0,0 +1,147 @@
+SectionLabel.propTypes = {
+  children: string,
+  dir: oneOf(['ltr', 'rtl']),
+  script: shape(scriptPropType),
```
Same as above, can we render a section label without `script` passed in? Should it be required?

sadickisaac replied:
@aacn500 I think script should be required. Failure to provide script causes the label to be out of line.
<img width="1471" alt="Screenshot 2019-05-17 at 12 06 26" src="https://user-images.githubusercontent.com/48688870/57916838-594d9a00-789c-11e9-937b-fbd3a2fe7528.png">

Commit:
Merge branch 'latest' into section-divider-nonflexbox

ghost commented:
The current implementation exports `SectionLabel` and `SectionLabelWithoutBar`. Have you looked into exporting `SectionLabel` and `SectionDivider` and have 
```jsx
<SectionDivider>
  <SectionLabel>
    Most Read
  </SectionLabel>
</SectionDivider>
```
just as you stated on your slack. IMO this will be a much better implementation. Glad to hear your thoughts on this.

sadickisaac commented on packages/components/psammead-section-label/README.md:
```diff
@@ -0,0 +1,80 @@
+
+The `@bbc/psammead-section-label` package exports two components - a Section Label, and a variant of the Section Label that does not feature a horizontal bar.
+
+It uses `@bbc/psammead-styles` for colours and font family and `@bbc/gel-foundations` for spacing and for GEL Typography implemented in Styled Components.
```
We might need to reword this 
```suggestion
For colours and font family it uses `@bbc/psammead-styles` and `@bbc/gel-foundations` for spacing and GEL Typography implemented in Styled Components.
```

andrew-nowak committed:
README update

andrew-nowak committed:
Merge branch 'section-label-opt1' into section-divider-nonflexbox

andrew-nowak committed:
Merge remote-tracking branch 'origin/latest' into section-divider-nonflexbox

andrew-nowak commented:
> The current implementation exports `SectionLabel` and `SectionLabelWithoutBar`. Have you looked into exporting `SectionLabel` and `SectionDivider` and have
> 
> ```js-jsx
> <SectionDivider>
>   <SectionLabel>
>     Most Read
>   </SectionLabel>
> </SectionDivider>
> ```
> 
> just as you stated on your slack. IMO this will be a much better implementation. Glad to hear your thoughts on this.

Good question! The short answer is that it doesn't really work without some pretty ugly logic.

1) To set the divider line to be vertically centered next to the label, I use `position: absolute;` and set `top` to half the text's line height. However, if the Label and the Divider are separate components, the Divider shouldn't require the `script` prop -- it shouldn't have text as a direct child. Even if we do accept the `script` prop, using those values without a Label would effectively increase the top margins and result in too large a spacing between sections. You technically *could* inspect the children, and if the child was a Label then you could use the values from its script prop, but that is definitely quite ugly.

2) The divider should control the spacing between sections, ie. sets a margin-top and margin-bottom, but if we export both a Divider and a Label and allow them to be used independently, how should the margins be applied? When trying to make this work, I ended up with

```css
*:not(${/* sc-selector */ SectionDivider}) > & {
  ${margins};
}
```

which IMO is worse than just using props to control whether the bar exists, and then being able to control this logic internally.

After some more discussion on slack, there was an agreement to move from the solution which you reviewed and use props to control the visibility of the horizontal bar, rather than a second export, which I have now pushed.

andrew-nowak committed:
Merge branch 'latest' into section-divider-nonflexbox

andrew-nowak committed:
Update packages/components/psammead-section-label/README.md

Co-Authored-By: sadickisaac <48688870+sadickisaac@users.noreply.github.com>

andrew-nowak committed:
update README

sadickisaac commented on packages/components/psammead-section-label/README.md:
```diff
@@ -0,0 +1,79 @@
+`npm install @bbc/psammead-section-label`
+
+## Props
+
```
You may also need to add prettier-ignore comment here.

sadickisaac commented on packages/components/psammead-section-label/README.md:
```diff
@@ -0,0 +1,79 @@
+| bar | boolean | no | `true` | `false` |
+| children | string | no | N/A | `'Most Read'` |
+| dir | string | no | `'ltr'` | `'rtl'` |
+| script | object | yes | N/A | `{ canon: { groupA: { fontSize: '28', lineHeight: '32',}, groupB: { fontSize: '32', lineHeight: '36' }, groupD: { fontSize: '44', lineHeight: '48' } }, trafalgar: { groupA: { fontSize: '20', lineHeight: '24' }, groupB: { fontSize: '24', lineHeight: '28' }, groupD: { fontSize: '32', lineHeight: '36' } } }` |
```
```suggestion
| script | object | yes | N/A | { canon: { groupA: { fontSize: '28', lineHeight: '32',}, groupB: { fontSize: '32', lineHeight: '36' }, groupD: { fontSize: '44', lineHeight: '48' } }, trafalgar: { groupA: { fontSize: '20', lineHeight: '24' }, groupB: { fontSize: '24', lineHeight: '28' }, groupD: { fontSize: '32', lineHeight: '36' } } } |
```
The back-ticks causes the script example to be displayed on one line, which becomes too long forcing a horizontal scroll on the browser.

Commit:
Merge branch 'latest' into section-divider-nonflexbox

andrew-nowak committed:
Update packages/components/psammead-section-label/README.md

Co-Authored-By: sadickisaac <48688870+sadickisaac@users.noreply.github.com>

andrew-nowak committed:
ignore prettier complaints in markdown table

andrew-nowak committed:
Merge branch 'latest' into section-divider-nonflexbox

sadickisaac commented on packages/components/psammead-section-label/README.md:
````diff
@@ -0,0 +1,80 @@
+## Usage
+
+```jsx
+import { SectionLabel } from '@bbc/psammead-section-label';
````
```suggestion
import SectionLabel from '@bbc/psammead-section-label';
```

sadickisaac commented on packages/components/psammead-section-label/README.md:
````diff
@@ -0,0 +1,80 @@
+Or, without a horizontal bar:
+
+```jsx
+import { SectionLabelWithoutBar } from '@bbc/psammead-section-label';
````
```suggestion
import SectionLabel from '@bbc/psammead-section-label';
```

sadickisaac commented on packages/components/psammead-section-label/CHANGELOG.md:
```diff
@@ -0,0 +1,6 @@
+<!-- prettier-ignore -->
+| Version | Description |
+|---------|-------------|
+| 0.1.0-alpha.0 | [PR#449](https://github.com/BBC/psammead/pull/449) Create initial package with Section Label components. |
```
```suggestion
| 0.1.0-alpha.1 | [PR#449](https://github.com/BBC/psammead/pull/449) Create initial package with Section Label components. |
```
Alpha versions usually start from `.1`

andrew-nowak replied:
do you have a reference for this? I don't see it mentioned on <https://semver.org>, and my understanding of the team's decision was to start at `.0`, and it was used for [story promo list](https://www.npmjs.com/package/@bbc/psammead-story-promo-list)

sadickisaac replied:
I can see it was indeed used for `story promo list`. I wasn't on the discussion on the `alpha` releases just ignore the suggestions. Ideally it should be either `0.1.0-alpha` or `0.1.0-alpha.1`.

sadickisaac commented on packages/components/psammead-section-label/package-lock.json:
```diff
@@ -0,0 +1,393 @@
+{
+  "name": "@bbc/psammead-section-label",
+  "version": "0.1.0-alpha.0",
```
```suggestion
  "version": "0.1.0-alpha.1",
```

sadickisaac commented on packages/components/psammead-section-label/package.json:
```diff
@@ -0,0 +1,43 @@
+{
+  "name": "@bbc/psammead-section-label",
+  "version": "0.1.0-alpha.0",
```
```suggestion
  "version": "0.1.0-alpha.1",
```

andrew-nowak committed:
prop-types uses `bool`, not `boolean`

Co-Authored-By: sadickisaac <48688870+sadickisaac@users.noreply.github.com>

andrew-nowak commented on packages/components/psammead-section-label/README.md:
````diff
@@ -0,0 +1,80 @@
+Or, without a section title:
+
+```jsx
+import { SectionLabel } from '@bbc/psammead-section-label';
````
```suggestion
import SectionLabel from '@bbc/psammead-section-label';
```

andrew-nowak committed:
Correct export type in README examples

Co-Authored-By: sadickisaac <48688870+sadickisaac@users.noreply.github.com>

andrew-nowak committed:
Merge branch 'latest' into section-divider-nonflexbox

DenisHdz reviewed (dismissed):
👍

andrew-nowak committed:
Merge branch 'latest' into section-divider-nonflexbox

andrew-nowak committed:
Require a title and enable passing in an id

passing in an id allows the container/other components to refer to the component with `aria-labelledby` or similar

andrew-nowak committed:
Update README with extra notes

sadickisaac commented on packages/components/psammead-section-label/src/index.jsx:
```diff
@@ -0,0 +1,114 @@
+};
+
+const SectionLabel = ({ bar, children: title, dir, labelId, script }) => (
+  <Wrapper script={title && script} bar={bar}>
```
I don't understand why we have `title && script` condition.

andrew-nowak replied:
Woops, I missed one. Well caught, thanks!

andrew-nowak committed:
bugfix; title is required and should be truthy

andrew-nowak commented:
There are remaining questions over how to display this component when there is not a `strapline` available. However, if required, that will be raised as a future issue. For now, we will assume that straplines are always provided.

andrew-nowak committed:
Merge branch 'latest' into section-divider-nonflexbox

dr3 reviewed (approved):
LGTM, great work 👍

david-boydell commented:
All good, thanks @aacn500, this is rfm.

jamesbhobbs committed:
Merge branch 'latest' into section-divider-nonflexbox
